src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -433,50 +445,58 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
         return CompletableFuture.completedFuture(0);
     }
 
-    private synchronized void spoolMessage()  {
-        if (spoolingFuture.get() == null || spoolingFuture.get().isCancelled()) {
-            spoolingFuture.set(ses.scheduleWithFixedDelay(this::spoolTask, 0, 5, TimeUnit.SECONDS));
+    protected CompletableFuture<Integer> publishSingleTask() throws InterruptedException {
+        final long id = spool.popId();
+        PublishRequest request = spool.getMessageById(id);
+        if (request == null) {
+            return null;
         }
-    }
-
-    /**
-     * Iterate the spooler queue to publish all the spooled message.
-     */
-    protected void spoolTask() {
-        try {
-            getConnection(false).connect().get();
-            List<CompletableFuture<?>> publishRequests = new ArrayList<>();
-            while (!Thread.currentThread().isInterrupted() && mqttOnline.get() && spool.getCurrentMessageCount() > 0) {
-                final long id = spool.popId();
-                PublishRequest request = spool.getMessageById(id);
-                if (request == null) {
-                    continue;
-                }
-
-                MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
-                publishRequests.add(getConnection(false).publish(m, request.getQos(), request.isRetain())
-                                .whenComplete((packetId, throwable) -> {
+        MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
+        return getConnection(false).publish(m, request.getQos(), request.isRetain())
+                .whenComplete((packetId, throwable) -> {
                     // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                     if (throwable == null) {
                         spool.removeMessageById(id);
                         logger.atDebug().kv("id", id).kv("topic", request.getTopic())
-                            .log("Successfully published message");
+                                .log("Successfully published message");
                     } else {
                         spool.addId(id);
                         logger.atError().log("Failed to publish the message via Spooler", throwable);
                     }
-                }));
+                });
+    }
+
+    /**
+     * Iterate the spooler queue to publish all the spooled message.
+     */
+    protected void publishAllTasks() {
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                getConnection(false).connect().get();
+                List<CompletableFuture<?>> publishRequests = new ArrayList<>();
+                while (waitMqttOnline.get()) {
```

Comment:
instead of waiting on a future, you can just quit the thread and save resources. Then when we're back online this will start executing again.